File: src/main/java/com/hazza/LeetCodeByJava/No62UniquePaths.java
package com.hazza.LeetCodeByJava;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/unique-paths/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-15
 * Time: 5:19 PM
 */
public class No62UniquePaths {
    public int uniquePaths(int m, int n) {
        if (m < 1 || n < 1) return 0;
        if (m == 1 || n == 1) return 1;
        int[] dp = new int[n + 1];
        dp[1] = 1;
        for (int i = 1 ; i <= m; i++)
            for (int j = 1; j <= n; j++)
                dp[j] = dp[j] + dp[j - 1];

        return dp[n];
    }
}

File: src/main/python/com/hazza/LeetCodeByPython/SegmentTwo/No62_Unique_Paths.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/12 20:53
# https://leetcode.com/problems/unique-paths/


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [0] * (n + 1)
        dp[1] = 1

        for _ in range(0, m):
            for i in range(1, n + 1):
                dp[i] = dp[i] + dp[i - 1]

        return dp[n]


if __name__ == '__main__':
    solution = Solution()

    assert solution.uniquePaths(3, 2) == 3
    assert solution.uniquePaths(7, 3) == 28
